src/main/java/jcifs/smb1/util/transport/Transport.java

```diff
@@ -0,0 +1,274 @@
+            } catch( InterruptedException ie ) {
+                throw new TransportException( ie );
+            } finally {
+                response_map.remove( request );
+            }
+    }
+    private void loop() {
+        while( thread == Thread.currentThread() ) {
+            try {
+                Request key = peekKey();
+                if (key == null)
+                    throw new IOException( "end of stream" );
+                synchronized (this) {
+                    Response response = (Response)response_map.get( key );
+                    if (response == null) {
+                        if (log.level >= 4)
+                            log.println( "Invalid key, skipping message" );
+                        doSkip();
+                    } else {
+                        doRecv( response );
+                        response.isReceived = true;
+                        notifyAll();
+                    }
+                }
+            } catch( Exception ex ) {
+                String msg = ex.getMessage();
+                boolean timeout = msg != null && msg.equals( "Read timed out" );
+                /* If just a timeout, try to disconnect gracefully
+                 */
+                boolean hard = timeout == false;
+
+                if (!timeout && log.level >= 3)
+                    ex.printStackTrace( log );
+
+                try {
+                    disconnect( hard );
+                } catch( IOException ioe ) {
+                    ioe.printStackTrace( log );
+                }
+            }
+        }
+    }
+
+    /* Build a connection. Only one thread will ever call this method at
+     * any one time. If this method throws an exception or the connect timeout
+     * expires an encapsulating TransportException will be thrown from connect
+     * and the transport will be in error.
+     */
+
+    protected abstract void doConnect() throws Exception;
+
+    /* Tear down a connection. If the hard parameter is true, the diconnection
+     * procedure should not initiate or wait for any outstanding requests on
+     * this transport.
+     */
+
+    protected abstract void doDisconnect( boolean hard ) throws IOException;
+
+    public synchronized void connect( long timeout ) throws TransportException {
+        try {
+            switch (state) {
+                case 0:
+                    break;
+                case 3:
+                    return; // already connected
+                case 4:
+                    state = 0;
+                    throw new TransportException( "Connection in error", te );
+                default:
+                    TransportException te = new TransportException( "Invalid state: " + state );
+                    state = 0;
+                    throw te;
+            }
+
+            state = 1;
+            te = null;
+            thread = new Thread( this, name );
+            thread.setDaemon( true );
+
+            synchronized (thread) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object thread synchronized on line 167 should not be re-assigned in the same synchronized block on line 179. After this re-assignment, you lose the reference to the synchronized object.